File: android/app/src/main/java/com/vatoo/erick/AppLanguageSupport.kt
package com.vatoo.erick

import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.runtime.staticCompositionLocalOf
import com.vatoo.erick.shared.ErickAppTranslations
import com.vatoo.erick.shared.KeyboardLanguage
import com.vatoo.erick.shared.KeyboardLanguageProfiles

val LocalAppLanguageKey = staticCompositionLocalOf { PreferencesManager.LANGUAGE_ENGLISH }
val LocalAppLocaleTag = staticCompositionLocalOf { KeyboardLanguageProfiles.localeTag(KeyboardLanguage.ENGLISH) }

@Composable
fun ProvideAppLanguage(
    preferencesManager: PreferencesManager,
    content: @Composable () -> Unit
) {
    val languageKey by preferencesManager.keyboardLanguage.collectAsState(initial = PreferencesManager.LANGUAGE_ENGLISH)
    val localeTag = remember(languageKey) {
        KeyboardLanguageProfiles.localeTag(languageKey.toKeyboardLanguage())
    }

    CompositionLocalProvider(
        LocalAppLanguageKey provides languageKey,
        LocalAppLocaleTag provides localeTag,
        content = content
    )
}

fun String.toKeyboardLanguage(): KeyboardLanguage = when (this) {
    PreferencesManager.LANGUAGE_SPANISH -> KeyboardLanguage.SPANISH
    PreferencesManager.LANGUAGE_PORTUGUESE -> KeyboardLanguage.PORTUGUESE
    PreferencesManager.LANGUAGE_FRENCH -> KeyboardLanguage.FRENCH
    PreferencesManager.LANGUAGE_GERMAN -> KeyboardLanguage.GERMAN
    PreferencesManager.LANGUAGE_ITALIAN -> KeyboardLanguage.ITALIAN
    PreferencesManager.LANGUAGE_NORWEGIAN_BOKMAL -> KeyboardLanguage.NORWEGIAN_BOKMAL
    PreferencesManager.LANGUAGE_DANISH -> KeyboardLanguage.DANISH
    PreferencesManager.LANGUAGE_SWEDISH -> KeyboardLanguage.SWEDISH
    PreferencesManager.LANGUAGE_FINNISH -> KeyboardLanguage.FINNISH
    else -> KeyboardLanguage.ENGLISH
}

fun erickText(languageKey: String, english: String): String {
    return ErickAppTranslations.text(languageKey.toKeyboardLanguage(), english)
}

fun englishLanguageDisplayName(languageKey: String): String = when (languageKey) {
    PreferencesManager.LANGUAGE_SPANISH -> "Spanish"
    PreferencesManager.LANGUAGE_PORTUGUESE -> "Portuguese"
    PreferencesManager.LANGUAGE_FRENCH -> "French"
    PreferencesManager.LANGUAGE_GERMAN -> "German"
    PreferencesManager.LANGUAGE_ITALIAN -> "Italian"
    PreferencesManager.LANGUAGE_NORWEGIAN_BOKMAL -> "Norwegian Bokmal"
    PreferencesManager.LANGUAGE_DANISH -> "Danish"
    PreferencesManager.LANGUAGE_SWEDISH -> "Swedish"
    PreferencesManager.LANGUAGE_FINNISH -> "Finnish"
    else -> "English"
}

File: ios/ERICK/ERICK/AppLanguageSupport.swift
import SwiftUI
import SharedKeyboard

let erickAppGroupDefaults = UserDefaults(suiteName: "group.com.vatoo.erick") ?? .standard

private struct ErickLanguageKey: EnvironmentKey {
    static let defaultValue = "english"
}

extension EnvironmentValues {
    var erickLanguageKey: String {
        get { self[ErickLanguageKey.self] }
        set { self[ErickLanguageKey.self] = newValue }
    }
}

func erickLocaleIdentifier(for keyboardLanguage: String) -> String {
    switch keyboardLanguage {
    case "spanish":
        return "es"
    case "portuguese":
        return "pt"
    case "french":
        return "fr"
    case "german":
        return "de"
    case "italian":
        return "it"
    case "norwegian_bokmal":
        return "nb"
    case "danish":
        return "da"
    case "swedish":
        return "sv"
    case "finnish":
        return "fi"
    default:
        return "en"
    }
}

func sharedKeyboardLanguage(for keyboardLanguage: String) -> KeyboardLanguage {
    switch keyboardLanguage {
    case "spanish":
        return .spanish
    case "portuguese":
        return .portuguese
    case "french":
        return .french
    case "german":
        return .german
    case "italian":
        return .italian
    case "norwegian_bokmal":
        return .norwegian_bokmal
    case "danish":
        return .danish
    case "swedish":
        return .swedish
    case "finnish":
        return .finnish
    default:
        return .english
    }
}

func erickText(_ english: String, languageKey: String) -> String {
    ErickAppTranslations.shared.text(language: sharedKeyboardLanguage(for: languageKey), english: english)
}

func recoverableEnglishTitle(for keyboardLanguage: String, english: String) -> String {
    let localized = erickText(english, languageKey: keyboardLanguage)
    return keyboardLanguage == "english" || localized.caseInsensitiveCompare(english) == .orderedSame
        ? localized
        : "\(localized) (\(english))"
}

func englishLanguageDisplayName(for keyboardLanguage: String) -> String {
    switch keyboardLanguage {
    case "spanish":
        return "Spanish"
    case "portuguese":
        return "Portuguese"
    case "french":
        return "French"
    case "german":
        return "German"
    case "italian":
        return "Italian"
    case "norwegian_bokmal":
        return "Norwegian Bokmal"
    case "danish":
        return "Danish"
    case "swedish":
        return "Swedish"
    case "finnish":
        return "Finnish"
    default:
        return "English"
    }
}